src/lib/formats/tzx_cas.cpp

```diff
@@ -291,10 +297,100 @@ static int tzx_cas_handle_block( int16_t **buffer, const uint8_t *bytes, int pau
 	return size;
 }
 
+static int tsx_msx_handle_block( int16_t **buffer, const uint8_t *bytes, int pause, int data_size, int pilot, int pilot_length, int bitcfg, int bytecfg, int bit0, int bit1)
+{
+	float t_scale = 1.44f;
+	int pilot_samples = tcycles_to_samplecount(pilot,t_scale);
+	int bit0_samples  = tcycles_to_samplecount(bit0 ,t_scale);
+	int bit1_samples  = tcycles_to_samplecount(bit1 ,t_scale);
+	int data_index;
+	int size = 0;
+	int bit1_pulses = (bitcfg & 0b00001111);
+	int bit0_pulses = (bitcfg & 0b11110000) >> 4;
+	int startvalue = (bytecfg & 0b00100000) >> 5;
+	int startbits = (bytecfg & 0b11000000) >> 6;
+	int stopvalue = (bytecfg & 0b00000100) >> 2;
+	int stopbits = (bytecfg & 0b00011000) >> 3;
+	
+
+
+
+	/* PILOT */
+	for ( ; pilot_length > 0; pilot_length--)
+	{
+		tzx_output_wave(buffer, pilot_samples);
+		size += pilot_samples;
+		toggle_wave_data();
+	}
+	
+	
+		/* data */
+	int start_bits = (startvalue & 0x01) ? bit1_samples : bit0_samples;
+	int stop_bits  = (stopvalue  & 0x01) ? bit1_samples : bit0_samples;
+	int multistart = startbits * ((startvalue & 0x01) ? bit1_pulses : bit0_pulses)/2;
+	int multistop  = stopbits  * ((stopvalue  & 0x01) ? bit1_pulses : bit0_pulses)/2;
+	for (data_index = 0; data_index < data_size; data_index++)
+	{
+		uint8_t byte = bytes[data_index];
+		
+		
+		
+		for (int startloop = 0; startloop < multistart; startloop++)
+		{
+			tzx_output_wave(buffer, start_bits);
+			size += start_bits;
+			toggle_wave_data();
+			tzx_output_wave(buffer, start_bits);
+			size += start_bits;
+			toggle_wave_data();
+		}
+		
+		int bits_to_go = 8;
+		for (;bits_to_go > 0; (byte>>=1) , bits_to_go--)
+		{
+			int bit_samples = (byte & 0x01) ? bit1_samples : bit0_samples;
+			for (int pulsesloop = 0; pulsesloop < ((byte & 0x01) ? bit1_pulses : bit0_pulses)/2; pulsesloop++)
+			{
+				tzx_output_wave(buffer, bit_samples);
+				size += bit_samples;
+				toggle_wave_data();
+				tzx_output_wave(buffer, bit_samples);
+				size += bit_samples;
+				toggle_wave_data();
+			}
+			
+			
```

Comment:
There’s no reason to have extra blank space here – it would make more sense to have space between the loops if anywhere.